[Step2] - 요금 조회

jongmin4943 opened the pull request:
안녕하세요 준우님 😃 
2단계 진행해 보았습니다. 요금 범위 계산부분을 책임연쇄패턴 + Range 로 엮을 수 있도록 구조를 짰는데 생각보다 너무 복잡하게 나왔습니다. 처음에는 각 클래스 별로 분리해서 구현했다가, 각 클래스들의 중복되는 코드들을 제거 하다 보니 하나의 클래스(?)가 되었습니다. 

만들면서도 살짝 망했는걸? 이런 느낌이 팍팍 들었지만, 한번 매질을 맞아야 더 좋은 구조가 떠오를 것 같습니다 😅
잘 부탁 드립니다 🙇‍♂️

jongmin4943 committed:
fix: SubwayMap graph 미리 생성, 메서드이름에 맞는 책임 부여 피드백 반영

jongmin4943 committed:
docs(README): 2단계 요구사항 `요금 조회` 정리

jongmin4943 committed:
fix: Path 인수테스트 요금 관련 추가

jongmin4943 committed:
feat: 거리별 요금 계산을 위한 `PathFareCalculator` 구성

jongmin4943 committed:
fix: CsvSource 를 통해 증가되는 경계 구간 테스트 추가

jongmin4943 committed:
fix: `PathFareCalculator` 구간별 조건 추가

jongmin4943 committed:
fix: `PathService` 에 PathFareCalculator 를 통한 요금 계산 로직 추가

jongmin4943 committed:
fix: 타입 깨지는 PathStepDef 수정

jongmin4943 committed:
fix: 요금 계산 구간별 책임연쇄패턴 적용

jongmin4943 committed:
refactor: 요금 계산 코드 리팩토링

jongmin4943 committed:
fix: 요금 계산 로직 오류 수정

jongmin4943 committed:
docs(README): 2단계 요구사항 체크

jongmin4943 commented on src/main/java/nextstep/path/application/FareChain.java:
```diff
@@ -0,0 +1,118 @@
+        return basePathFare.calculate(distance);
+    }
+
+    private static class PathFare {
```
하나의 클래스가 되면서 시작점인 Base 를 제외하고는 외부로 노출될 필요가 없다고 느껴 Base Class 의 `static inner class` 구조로 짜 보았습니다. 두가지 장점을 가져가 보려했는데요 :)
- 어떤 클래스가 한 클래스에서만 유용하게 쓰인다면 물리적으로 하나의 클래스로 묶을 수 있다.
- private 수준에서도 내부클래스는 외부클래스 맴버 또는 메서드에 접근할 수 있기 때문에 캡슐화에 용이하다.

`HashMap` 내부의 `EntrySet Inner class` 구조를 참고해보았습니다 😃 어떻게 생각하시나요?

junwoochoi replied:
외부 노출을 고의적으로 막으셔서 캡슐화를 잘해주신 것 같네요! 👍 
다만, PathFare가 만약 인터페이스로 분리되어 구현체가 여러개가 된다면 분리해주시는 방향도 좋을 것 같습니다!

jongmin4943 commented on src/main/java/nextstep/path/application/FareChain.java:
```diff
@@ -0,0 +1,118 @@
+            this.fareInterval = fareInterval;
+        }
+
+        private long calculate(final int distance) {
+            final int targetDistance = getTargetDistance(distance);
+
+            int extraFare = 0;
+            if (targetDistance > standardDistance) {
+                extraFare += calculateOverFare(targetDistance - standardDistance);
+            }
+
+            return extraFare + calculateNext(distance);
+        }
```
복잡도가 올라간 지점이라고 생각되는 부분인데요 🤔
체이닝 방식으로 한개씩 차례대로 돌면서 추가비용을 계산하는데, `현재 범위부터 다음 범위까지만 요금 계산`하고, 다음으로 넘기는 로직이에요. 이 부분이 코드의 복잡도를 높이고 있다 생각해요 😥
이 부분 때문에 수 많이 바꿨다가, 지웠다가 하면서 이리저리 헤집어 봤지만 영 좋은 구조가 떠오르지 않더라구요. 준우님은 어떻게 생각하시나요?

junwoochoi replied:
책임연쇄패턴을 생각하고 코드를 잘 작성해주신 것 같아요! :)
다만, 현재 PathFare에는 체인을 관리하는 로직과 추가 요금을 계산하는 로직 + 전체 요금을 계산하는 로직이 다 같이 있어서 복잡도가 높다고 느껴지는걸로 보여요.
해당 부분을 인터페이스로 분리하고 각각의 요금을 계산하는 부분을 체이닝해서 별도 클래스로 분리하면 조금 더 깔끔한 코드로 정리가 되지 않을까요? 👀

junwoochoi commented on README.md:
```diff
@@ -43,3 +43,26 @@ And 지하철 노선에 지하철역이 등록되어있음<br>
 Then 최소 시간 기준 경로를 응답<br>
 And 총 거리와 소요 시간을 함께 응답함
+
+## 2단계 - 요금 조회
+### 요구사항
+#### 경로 조회 시 요금 정보 포함
+- [x] 경로 조회 결과에 요금 정보를 포함
+  - 기본운임(10km 이내) : 기본운임 1,250원
+  - 이용 거리초과 시 추가운임 부과
+    - 10km 초과 ~ 50km 이하 - 5km 마다 100원 추가
+    - 50km 초과 - 8km 마다 100원 추가
+    > 9km = 1250원<br>
+    12km = 10km + 2km = 1350원<br>
+    16km = 10km + 6km = 1450원
+
+### 인수 조건
+- Feature: 지하철 경로 검색
+>  Scenario: 두 역의 최단 거리 경로를 조회 <br>
+Given 지하철역이 등록되어있음 <br>
+And 지하철 노선이 등록되어있음 <br>
+And 지하철 노선에 지하철역이 등록되어있음 <br>
+When 출발역에서 도착역까지의 최단 거리 경로 조회를 요청 <br>
+Then 최단 거리 경로를 응답 <br>
+And 총 거리와 소요 시간을 함께 응답함 <br>
+And 지하철 이용 요금도 함께 응답함
```
요구사항 정리 👍

junwoochoi commented on src/main/java/nextstep/path/domain/SubwayMap.java:
```diff
@@ -3,34 +3,25 @@
-        }
-
-        return new DijkstraShortestPath<>(graph);
+        this.graphCache = Arrays.stream(PathType.values())
```
👍 👍 👍

junwoochoi commented on src/test/java/nextstep/cucumber/steps/PathStepDef.java:
```diff
@@ -26,13 +26,19 @@ public PathStepDef() {
-                softly.assertThat(context.response.jsonPath().getInt("duration")).isEqualTo(duration);
-            });
-        });
+        Then("{string} 경로와 거리 {int}, 소요시간 {int}, 요금 {long}원이 조회된다",
```
요금관련해 변경된 부분을 인수테스트부분에도 변경 반영해주셨네요 👍

junwoochoi commented on src/test/java/nextstep/path/application/FareChainTest.java:
```diff
@@ -0,0 +1,63 @@
+     * 25 1750
+     * 30 1850
+     */
+    @Test
```
이런 부분은 주석에 적어주신 부분을 `@ParameterizedTest`를 활용해서 다 같이 테스트해봐도 좋겠네요 :)

junwoochoi commented on src/test/java/nextstep/path/application/FareChainTest.java:
```diff
@@ -0,0 +1,63 @@
+        assertThat(calculated).isEqualTo(1850);
+    }
+
+    @Test
+    @DisplayName("두번째 범위 시작점이 첫번째 범위 시작점보다 작을 수 없다")
+    void standardDistanceLtPrevTest() {
+       assertThatThrownBy(() -> new FareChain()
+               .nextRange(5, 2)
+               .nextRange(4, 1))
+               .isInstanceOf(FareApplyingException.class)
+               .hasMessageContaining("standardDistance must be grater than previous standardDistance");
+    }
+    @Test
+    @DisplayName("구간 요금 반복 범위가 1 보다 작을 수 없다")
+    void fareIntervalLtOneDistance() {
+       assertThatThrownBy(() -> new FareChain()
+               .nextRange(1, 0))
+               .isInstanceOf(FareApplyingException.class)
+               .hasMessageContaining("fareInterval must be grater than");
+    }
+    @Test
+    @DisplayName("범위 시작점이 1 보다 작을 수 없다")
```
단위테스트에서 예외케이스도 꼼꼼히 확인하신 부분 좋습니다! 💯

junwoochoi reviewed (changes requested):
안녕하세요 종민님 :)
2단계 미션도 요구사항에 맞춰 잘구현해주시고, 단위테스트+인수테스트도 꼼꼼히 챙겨주셨습니다!
3단계로 넘어가기전에 몇가지 간단한 코멘트 달아보았으니 한번만 더 확인부탁드려요!


jongmin4943 committed:
fix: FareChainTest `@ParameterizedTest` 적용

jongmin4943 committed:
fix: FareChain 책임 분리

- FareChain -> 체인을 관리
- 추가 요금 계산 -> FirstExtraFareHandler, SecondExtraFareHandler
- 전체 요금 계산 -> BaseFareHandler

jongmin4943 commented:
다시 기존에 짰던 구조로 돌려보았습니다. 
- FareChain -> 체인을 관리
- 추가 요금 계산 -> FirstExtraFareHandler, SecondExtraFareHandler
- 전체 요금 계산 -> BaseFareHandler

중복되는 로직을 한군데 모으다보니 `PathFare` 내부로 모였었는데, 그 부분이 오히려 클래스의 복잡도를 높이고 책임이 모이게 되었었군요. 말씀해주신 구조가 이 구조가 맞을까요?

junwoochoi commented on src/test/java/nextstep/path/application/FareChainTest.java:
```diff
@@ -13,51 +15,55 @@ class FareChainTest {
-     * 30 1850
      */
-    @Test
+    @ParameterizedTest
```
👍

jongmin4943 replied:
> 정답은 없는 부분이니 많은 의견 주시면서 더좋은 구조를 찾아봐도 좋을것같습니다.

저도 이런 저런 시도를 많이 해보긴 했는데 이 형태가 가장 직관적인 형태라고 생각해요 ㅎㅎ
코드의 중복을 줄여보기 위해 하나로 모았었는데, 사실 각자의 정책은 각자가 가지고 구현 하는게 맞다고 생각하고,
단지 중복되었던것은, 그때 마침 계산하는 로직이 우연히 같았다고 생각이 드네요. 😃

junwoochoi commented on src/main/java/nextstep/path/application/BaseFareHandler.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public long calculate(final int distance) {
+        return BASE_FARE + calculateNext(distance);
```
calculateNext를 호출하는걸 잊지않도록 인터페이스나 abstract 클래스단에서 calculateNext는 호출하고 구현체는 계산하는 부분만 구현하도록하면 더 안전한 코드가 될거같아요~

jongmin4943 replied:
그렇네요. 각자 자신의 Fare 만 계산하게 하고 다음으로 넘기는걸 직접 호출하게 되면 문제가 될 수 있겠네요. 반영하겠습니다!

junwoochoi reviewed (approved):
안녕하세요! 제가 말씀드린 구조와 더 가깝게 잘 수정해주셨습니다! ㅎㅎ 다만 정답은 없는 부분이니 많은 의견 주시면서 더좋은 구조를 찾아봐도 좋을것같습니다 😃 
한가지 코멘트 달아보았으니 확인 부탁드리고 다음 미션도 이어서 화이팅해주세요 🚀
